nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java

```diff
@@ -0,0 +1,285 @@
+            final EventBatch eventBatch = getEventBatch(eventStream, pollDuration, pollDurationTimeUnit);
+
+            if (eventBatch != null && eventBatch.getEvents() != null) {
+                if (eventBatch.getEvents().length > 0) {
+                    List<FlowFile> flowFiles = new ArrayList<>(eventBatch.getEvents().length);
+                    for (Event e : eventBatch.getEvents()) {
+                        if (toProcessEvent(context, e)) {
+                            getLogger().debug("Creating flow file for event.");
+                            FlowFile flowFile = session.create();
+                            flowFile = session.putAttribute(flowFile, EventAttributes.MIME_TYPE, "application/json");
+                            flowFile = session.putAttribute(flowFile, EventAttributes.EVENT_TYPE, e.getEventType().name());
+                            flowFile = session.putAttribute(flowFile, EventAttributes.EVENT_PATH, getPath(e));
+                            flowFile = session.write(flowFile, new OutputStreamCallback() {
+                                @Override
+                                public void process(OutputStream out) throws IOException {
+                                    out.write(OBJECT_MAPPER.writeValueAsBytes(e));
+                                }
+                            });
+                            flowFiles.add(flowFile);
+                        }
+                    }
+
+                    getLogger().debug("Transferring all event flow files.");
+                    session.transfer(flowFiles, REL_SUCCESS);
+                }
+
+                lastTxId = eventBatch.getTxid();
+            }
+        } catch (IOException | InterruptedException e) {
+            getLogger().error("Unable to get notification information: {}", new Object[]{e});
+            throw new ProcessException(e);
+        } catch (MissingEventsException e) {
+            // set lastTxId to -1 and update state. This may cause events not to be processed. The reason this exception is thrown is described in the
+            // org.apache.hadoop.hdfs.client.HdfsAdmin#getInotifyEventStrea API. It suggests tuning a couple parameters if this API is used.
+            lastTxId = -1L;
+            getLogger().error("Unable to get notification information. Setting transaction id to -1. This may cause some events to get missed. " +
+                    "Please see javadoc for org.apache.hadoop.hdfs.client.HdfsAdmin#getInotifyEventStream: {}", new Object[]{e});
+        }
+
+        updateClusterStateForTxId(stateManager);
+    }
+
+    private EventBatch getEventBatch(DFSInotifyEventInputStream eventStream, long duration, TimeUnit timeUnit) throws IOException, InterruptedException, MissingEventsException {
+        // According to the inotify API we should retry a few times if poll throws an IOException.
+        // Please see org.apache.hadoop.hdfs.DFSInotifyEventInputStream#poll for documentation.
+        int i = 0;
```

Comment:
Would it make sense to have this "retry number" as a user-defined property? just asking.